### PR TITLE
[YUNIKORN-2855] Rethink the call for update metrics with nil resource

### PR DESCRIPTION
**Pull request by zhuqi-lucas**
### What is this PR for?
We will not update the nil resource to resource metrics, we need to fix it.

### What type of PR is it?
* [ ] - Bug Fix
* [x] - Improvement
* [ ] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
* [ ] - Task

### What is the Jira issue?
* Open an issue on Jira https://issues.apache.org/jira/browse/YUNIKORN-2855
* Put link here, and add [YUNIKORN-*Jira number*] in PR title, eg. `[YUNIKORN-2] Gang scheduling interface parameters`

### Questions:
* [ ] - The licenses files need update.
* [ ] - There is breaking changes for older versions.
* [ ] - It needs documentation.

### CODE REVIEW

**Review by craigcondit — CHANGES_REQUESTED** [review]
I'm -1 on this approach. This adds significant complexity to the code and requires all callers to do the calculations... make the metrics functions handle the nil case properly themselves.

**Comment by zhuqi-lucas** [comment]
Thanks @craigcondit for review, i addressed the code to a more reasonable way and setting zero resource for nil update case.

**Review by craigcondit — COMMENTED** [review]
Maybe I'm missing something entirely... Why are we setting arbitrary resources to zero here? What is it we're trying to accomplish?

**Comment by zhuqi-lucas** [comment]
@craigcondit , we support setting maxResource/ guaranteed resource to nil, but the metrics update is wrong because we don't update it from previous value update to zero/nil.

**Comment by craigcondit** [comment]
Then it seems the proper solution is to retrieve the previous value of the metric and update the values there (an "unprune" so to speak). Missing values n the new map should be set to zero. Setting arbitrary resources is just wrong. Doing it this way also has the advantage of ensuring that any resource types that were emitted by the metric previously are still emitted, but with zero instead of missing.

**Comment by zhuqi-lucas** [comment]
Thanks, @craigcondit i like the idea unprune, i will address this way.

**Review by craigcondit — CHANGES_REQUESTED** [review]
Just... no. You already have the metric in the collector itself. Read it out to get the list of values that are previously set and build a new metric using a combination of the old and new values. There should be no code changes outside the collector at all.

**Comment by zhuqi-lucas** [comment]
I see... @craigcondit let me change my code.

**Comment by zhuqi-lucas** [comment]
Addressed in latest PR:

1. Getting the existed metrics.
2. Setting those field to zero.

File: pkg/scheduler/objects/queue.go [hunk 1]
```diff
@@ -1659,17 +1659,21 @@ func (sq *Queue) SupportTaskGroup() bool {
 func (sq *Queue) updateGuaranteedResourceMetrics() {
 	if sq.guaranteedResource != nil {
 		for k, v := range sq.guaranteedResource.Resources {
```
Comment (craigcondit) [inline pkg/scheduler/objects/queue.go:1661]: There's no reason for any of this logic to be here. Encapsulate all of it within pkg/metrics/queue.go.

File: pkg/scheduler/objects/queue.go [hunk 2]
```diff
@@ -1659,17 +1659,21 @@ func (sq *Queue) SupportTaskGroup() bool {
...
+	} else {
+		metrics.GetQueueMetrics(sq.QueuePath).SetQueueNilResourceMetrics(metrics.QueueGuaranteed)
 	}
 }
 
 // updateMaxResourceMetrics updates max resource metrics.
 func (sq *Queue) updateMaxResourceMetrics() {
 	if sq.maxResource != nil {
 		for k, v := range sq.maxResource.Resources {
-			metrics.GetQueueMetrics(sq.QueuePath).SetQueueMaxResourceMetrics(k, float64(v))
```
Comment (craigcondit) [inline pkg/scheduler/objects/queue.go:1671]: Same - move logic into pkg/metrics/queue.go.

File: pkg/metrics/queue.go [hunk 3]
```diff
@@ -301,12 +301,46 @@ func (m *QueueMetrics) AddReleasedContainers(value int) {
...
 
-func (m *QueueMetrics) SetQueueGuaranteedResourceMetrics(resourceName string, value float64) {
-	m.setQueueResource(QueueGuaranteed, resourceName, value)
+func (m *QueueMetrics) SetQueueResourceMetrics(state string, resourceName string, value float64) {
+	m.setQueueResource(state, resourceName, value)
 }
 
-func (m *QueueMetrics) SetQueueMaxResourceMetrics(resourceName string, value float64) {
-	m.setQueueResource(QueueMax, resourceName, value)
+func (m *QueueMetrics) SetQueueNilResourceMetrics(state string) {
```
Comment (craigcondit) [inline pkg/metrics/queue.go:308]: This should be made an internal function once logic is moved here from scheduler/objects/queue.go.

File: pkg/metrics/queue.go [hunk 4]
```diff
@@ -301,12 +301,46 @@ func (m *QueueMetrics) AddReleasedContainers(value int) {
...
+}
+
+func (m *QueueMetrics) GetAllQueueStateMetrics(state string) (map[string]float64, error) {
+	metrics := make(map[string]float64)
+	metricFamilies, err := prometheus.DefaultGatherer.Gather()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, metricFamily := range metricFamilies {
```
Comment (craigcondit) [inline pkg/metrics/queue.go:327]: Can this be simplified? This seems... extreme. Perhaps we can simply keep (as part of the QueueMetrics object) a list of known resource types for the current queue. Then we can pass in all the resources types at once into SetQueueResourceMetrics(). Iterate both old and new types. Old types that don't exist in the new collection get emitted with zero; new types get added to the known list. Values in the new collection get emitted as-is. Be sure to us appropriate locking.

This allows a much more efficient lookup that doesn't involve traversing the entirety of the metrics, and provides a much more convenient interface for callers.

**Review by craigcondit — CHANGES_REQUESTED** [review]
The code is still too complex. See review comments.

File: pkg/metrics/queue.go (hunk 4 above)
Comment (zhuqi-lucas) [reply pkg/metrics/queue.go:327]: Good idea, i will address this in new PR.

File: pkg/scheduler/objects/queue.go (hunk 1 above)
Comment (zhuqi-lucas) [reply pkg/scheduler/objects/queue.go]: Addressed in latest PR, thanks.

File: pkg/metrics/queue.go (hunk 4 above)
Comment (zhuqi-lucas) [reply pkg/metrics/queue.go]: It seems we have a lock for metric level, but not queue level:
```
func init() {
	once.Do(func() {
		m = &Metrics{
			scheduler: InitSchedulerMetrics(),
			queues:    make(map[string]*QueueMetrics),
			event:     initEventMetrics(),
			lock:      locking.RWMutex{},
			runtime:   initRuntimeMetrics(),
		}
	})
}
```

Do we need to create a queue level lock to lock the knownResourceType?

File: pkg/metrics/queue.go [hunk 5]
```diff
@@ -57,6 +58,8 @@ type QueueMetrics struct {
 	resourceMetricsLabel *prometheus.GaugeVec
 	// Deprecated - To be removed in 1.7.0. Replaced with queue label Metrics
 	resourceMetricsSubsystem *prometheus.GaugeVec
+	// Track known resource types.
+	knownResourceTypes map[string]struct{}
```
Comment (craigcondit) [inline pkg/metrics/queue.go:62]: Need a lock here to guard access to knownResourceTypes.

File: pkg/metrics/queue.go [hunk 6]
```diff
@@ -146,6 +150,7 @@ func (m *QueueMetrics) Reset() {
 	m.appMetricsSubsystem.Reset()
 	m.resourceMetricsLabel.Reset()
 	m.resourceMetricsSubsystem.Reset()
+	m.knownResourceTypes = make(map[string]struct{})
```
Comment (craigcondit) [inline pkg/metrics/queue.go:153]: Guard this with a lock (see previous comment).

File: pkg/metrics/queue.go [hunk 7]
```diff
@@ -301,12 +306,20 @@ func (m *QueueMetrics) AddReleasedContainers(value int) {
...
 
-func (m *QueueMetrics) SetQueueGuaranteedResourceMetrics(resourceName string, value float64) {
-	m.setQueueResource(QueueGuaranteed, resourceName, value)
-}
+func (m *QueueMetrics) UpdateQueueResourceMetrics(state string, newResources map[string]resources.Quantity) {
+	// Iterate over new resource types and set their values.
+	for resourceName, value := range newResources {
+		m.setQueueResource(state, resourceName, float64(value))
+		// Add new resources to the known list
+		m.knownResourceTypes[resourceName] = struct{}{}
```
Comment (craigcondit) [inline pkg/metrics/queue.go:314]: Guard this with a lock (see previous comment).

**Review by craigcondit — CHANGES_REQUESTED** [review]
Add locking per comments.

File: pkg/metrics/queue.go (hunk 4 above)
Comment (craigcondit) [reply pkg/metrics/queue.go]: The individual metric instance needs a lock (see review comments).

File: pkg/metrics/queue.go (hunk 5 above)
Comment (zhuqi-lucas) [reply pkg/metrics/queue.go:63]: Addressed in latest PR, thanks!

File: pkg/metrics/queue.go (hunk 6 above)
Comment (zhuqi-lucas) [reply pkg/metrics/queue.go:157]: Addressed in latest PR, thanks!

File: pkg/metrics/queue.go (hunk 7 above)
Comment (zhuqi-lucas) [reply pkg/metrics/queue.go:320]: Addressed in latest PR, thanks!

File: pkg/metrics/queue.go [hunk 8]
```diff
@@ -58,8 +59,9 @@ type QueueMetrics struct {
 	resourceMetricsLabel *prometheus.GaugeVec
 	// Deprecated - To be removed in 1.7.0. Replaced with queue label Metrics
 	resourceMetricsSubsystem *prometheus.GaugeVec
-	// Track known resource types.
+	// Track known resource types
 	knownResourceTypes map[string]struct{}
+	lock               locking.RWMutex
```
Comment (craigcondit) [inline pkg/metrics/queue.go:64]: This doesn't need to be an RWMutex as we never lock for reads.

**Review by craigcondit — CHANGES_REQUESTED** [review]
Change lock type.

File: pkg/metrics/queue.go (hunk 8 above)
Comment (zhuqi-lucas) [reply pkg/metrics/queue.go]: Good catch, addressed in latest PR, thanks!

**Review by craigcondit — APPROVED** [review]
+1 LGTM.

**Pull request closed without merging**